### PR TITLE
feat: add reason parameter in disconnect events

### PR DESCRIPTION
**Pull request by VivekPipaliya23**
## Proposed Changes

Add Reason parameter in `on_client_offline` and `on_client_gone` events. The `Reason` parameter is used to provide additional information about the client's disconnection reason.

## Types of Changes

What types of changes does your code introduce to this project?
_Put an `x` in the boxes that apply_

- [ ] Bugfix (non-breaking change which fixes issue #XXXX)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation (correction or otherwise)
- [ ] Cosmetics (whitespace, styles...)
- [ ] DevOps (Build scripts, pipelines...)

### CODE REVIEW

File: apps/vmq_proto/include/on_client_gone_pb.hrl [hunk 1]
```diff
@@ -31,6 +31,12 @@
             | 'REASON_MQTT_CLIENT_DISCONNECT'
             | 'REASON_RECEIVE_MAX_EXCEEDED'
             | 'REASON_PROTOCOL_ERROR'
+            | 'REASON_PUBLISH_AUTH_ERROR'
+            | 'REASON_INVALID_PUBREC_ERROR'
+            | 'REASON_INVALID_PUBCOMP_ERROR'
+            | 'REASON_UNEXPECTED_FRAME_TYPE'
+            | 'REASON_EXIT_SIGNAL_RECIVED'
```
Comment (dhruvjain99) [inline apps/vmq_proto/include/on_client_gone_pb.hrl:38]: RECEIVED -> typo

File: apps/vmq_server/src/vmq_mqtt_fsm_util.erl [hunk 2]
```diff
@@ -160,5 +162,12 @@ terminate_proto_reason(Reason) ->
         mqtt_client_disconnect -> 'REASON_MQTT_CLIENT_DISCONNECT';
         receive_max_exceeded -> 'REASON_RECEIVE_MAX_EXCEEDED';
         protocol_error -> 'REASON_PROTOCOL_ERROR';
+        publish_not_authorized_3_1_1 -> 'REASON_PUBLISH_AUTH_ERROR';
```
Comment (dhruvjain99) [inline apps/vmq_server/src/vmq_mqtt_fsm_util.erl:165]: Create MACROS for these constants.

File: apps/vmq_server/src/vmq_mqtt_fsm_util.erl [hunk 3]
```diff
@@ -160,5 +162,12 @@ terminate_proto_reason(Reason) ->
...
+        invalid_pubrec_error -> 'REASON_INVALID_PUBREC_ERROR';
+        invalid_pubcomp_error -> 'REASON_INVALID_PUBCOMP_ERROR';
+        unexpected_frame_type -> 'REASON_UNEXPECTED_FRAME_TYPE';
+        exit_signal_recived -> 'REASON_EXIT_SIGNAL_RECIVED';
+        tcp_closed -> 'REASON_TCP_CLOSED';
+        normal -> 'REASON_NORMAL_DISCONNECT';
         _ -> 'REASON_UNSPECIFIED'
```
Comment (dhruvjain99) [inline apps/vmq_server/src/vmq_mqtt_fsm_util.erl:172]: KEEPALIVE REASON MISSING.

File: apps/vmq_server/src/vmq_mqtt_pre_init.erl [hunk 4]
```diff
@@ -168,7 +168,7 @@ get_protocol_info(_) ->
 
 -spec msg_in('close_timeout' | 'disconnect' | {'disconnect', 'normal_disconnect'}, _) ->
```
Comment (dhruvjain99) [inline apps/vmq_server/src/vmq_mqtt_pre_init.erl:169]: Change spec accordingly.

File: apps/vmq_server/src/vmq_ranch.erl [hunk 5]
```diff
@@ -259,7 +259,7 @@ handle_message({Proto, _, Data}, #st{proto_tag = {Proto, _, _}, fsm_mod = FsmMod
     end;
 handle_message({ProtoClosed, _}, #st{proto_tag = {_, ProtoClosed, _}, fsm_mod = FsmMod} = State) ->
     %% we regard a tcp_closed as 'normal'
-    _ = FsmMod:msg_in({disconnect, ?NORMAL_DISCONNECT}, State#st.fsm_state),
+    _ = FsmMod:msg_in({disconnect, tcp_closed}, State#st.fsm_state),
```
Comment (dhruvjain99) [inline apps/vmq_server/src/vmq_ranch.erl:262]: USE MACROS

File: apps/vmq_server/src/vmq_mqtt_fsm_util.erl [hunk 6]
```diff
@@ -144,6 +144,8 @@ terminate_reason(?DISCONNECT_MIGRATION) -> normal;
 terminate_reason(?NORMAL_DISCONNECT) -> normal;
 terminate_reason(?SESSION_TAKEN_OVER) -> normal;
 terminate_reason(?REMOTE_SESSION_TAKEN_OVER) -> normal;
+terminate_reason(?INVALID_PUBREC_ERROR) -> normal;
+terminate_reason(?INVALID_PUBCOMP_ERROR) -> normal;
```
Comment (dhruvjain99) [inline apps/vmq_server/src/vmq_mqtt_fsm_util.erl:148]: Map new REASONS to normal. Also where you have changed normal to specific reasons, normal mapping must be done here.

**Review by dhruvjain99 — CHANGES_REQUESTED** [review]

File: apps/vmq_proto/include/on_client_gone_pb.hrl (hunk 1 above)
Comment (VivekPipaliya23) [reply apps/vmq_proto/include/on_client_gone_pb.hrl:38]: Done

File: apps/vmq_server/src/vmq_mqtt_fsm_util.erl (hunk 3 above)
Comment (VivekPipaliya23) [reply apps/vmq_server/src/vmq_mqtt_fsm_util.erl:172]: It is already there on line 158.

File: apps/vmq_server/src/vmq_ranch.erl (hunk 5 above)
Comment (VivekPipaliya23) [reply apps/vmq_server/src/vmq_ranch.erl:262]: DONE

File: apps/vmq_server/src/vmq_mqtt_fsm_util.erl (hunk 2 above)
Comment (VivekPipaliya23) [reply apps/vmq_server/src/vmq_mqtt_fsm_util.erl:165]: Done

File: apps/vmq_server/src/vmq_mqtt_pre_init.erl (hunk 4 above)
Comment (VivekPipaliya23) [reply apps/vmq_server/src/vmq_mqtt_pre_init.erl:169]: Done

File: apps/vmq_server/src/vmq_mqtt_fsm_util.erl (hunk 6 above)
Comment (VivekPipaliya23) [reply apps/vmq_server/src/vmq_mqtt_fsm_util.erl:148]: Done

File: apps/vmq_server/src/vmq_mqtt_fsm_util.erl (hunk 3 above)
Comment (VivekPipaliya23) [reply apps/vmq_server/src/vmq_mqtt_fsm_util.erl:174]: Now on 160

File: apps/vmq_proto/include/on_client_gone_pb.hrl (hunk 1 above)
Comment (dhruvjain99) [reply apps/vmq_proto/include/on_client_gone_pb.hrl:38]: Where is this done? I can't see in the changes.

Comment (VivekPipaliya23) [reply apps/vmq_proto/include/on_client_gone_pb.hrl]: It should be there now.

**Review by dhruvjain99 — CHANGES_REQUESTED** [review]
Let's add metrics for tracking different reasons as well. Please resolve the comments by adding new commits. And I expect each commit to resolve only related comments. Changes for resolving unrelated comments must not be in the same commit.

File: apps/vmq_proto/proto/on_client_gone.proto [hunk 7]
```diff
@@ -1,6 +1,7 @@
 syntax = "proto3";
 
 import "google/protobuf/timestamp.proto";
+import "disconnect_reason.proto";
 
 option go_package = "source.golabs.io/courier/apis-go/eventssidecar/v1";
```
[the remark this answers is not in the document]
Comment (VivekPipaliya23) [reply apps/vmq_proto/proto/on_client_gone.proto]: Done

File: apps/vmq_proto/proto/disconnect_reason.proto [hunk 8]
```diff
@@ -0,0 +1,26 @@
+syntax = "proto3";
+
+option go_package = "source.golabs.io/courier/apis-go/eventssidecar/v1";
```
[the remark this answers is not in the document]
Comment (VivekPipaliya23) [reply apps/vmq_proto/proto/disconnect_reason.proto]: Done

File: apps/vmq_diversity/test/vmq_diversity_plugin_SUITE.erl [hunk 9]
```diff
@@ -312,3 +312,4 @@ payload() -> <<"hello world">>.
 subopts() ->
     #{rap => true,
       no_local => false}.
+reason() -> normal_disconnect.
```
[the remark this answers is not in the document]
Comment (VivekPipaliya23) [reply apps/vmq_diversity/test/vmq_diversity_plugin_SUITE.erl]: Done

File: apps/vmq_diversity/test/vmq_diversity_auth_cache_SUITE.erl [hunk 10]
```diff
@@ -163,3 +163,4 @@ allowed_subscriber_id() ->
 username() -> <<"test-user">>.
 password() -> <<"test-password">>.
 payload() -> <<"hello world">>.
+reason() -> normal_disconnect.
```
[the remark this answers is not in the document]
Comment (VivekPipaliya23) [reply apps/vmq_diversity/test/vmq_diversity_auth_cache_SUITE.erl]: Done

**Review by dhruvjain99 — CHANGES_REQUESTED** [review]

File: apps/vmq_server/src/vmq_metrics.erl [hunk 11]
```diff
@@ -1818,6 +1819,40 @@ rcn_to_str(RNC) ->
...
+            ?REASON_NORMAL,
+            ?REASON_UNSPECIFIED
+        ],
+    [
+        m(
+            counter,
+            [{mqtt_version, "4"}, {reason_code, rcn_to_str(RCN)}],
+            RCN,
+            mqtt_disconnect,
+            <<"The number of DISCONNECT packets received.">>
```
[the remark this answers is not in the document]
Comment (VivekPipaliya23) [reply apps/vmq_server/src/vmq_metrics.erl]: Done

File: apps/vmq_server/src/vmq_metrics.hrl [hunk 12]
```diff
@@ -137,3 +137,22 @@
 -define(RETRY, false).
 -define(QOS1_NON_RETRY_DROPPED, qos1_non_retry_dropped).
 -define(QOS1_NON_PERSISTENCE_DROPPED, qos1_non_persistence_dropped).
+-define(REASON_NOT_AUTHORIZED, not_authorized).
```
[the remark this answers is not in the document]
Comment (VivekPipaliya23) [reply apps/vmq_server/src/vmq_metrics.hrl]: Updated according to proto

File: apps/vmq_server/src/vmq_metrics.erl [hunk 13]
```diff
@@ -1818,6 +1819,40 @@ rcn_to_str(RNC) ->
...
+            ?REASON_EXIT_SIGNAL_RECEIVED,
+            ?REASON_TCP_CLOSED,
+            ?REASON_NORMAL,
+            ?REASON_UNSPECIFIED
+        ],
+    [
+        m(
+            counter,
+            [{mqtt_version, "4"}, {reason_code, rcn_to_str(RCN)}],
+            RCN,
```
[the remark this answers is not in the document]
Comment (VivekPipaliya23) [reply apps/vmq_server/src/vmq_metrics.erl]: done

File: apps/vmq_server/src/vmq_mqtt_fsm_util.erl [hunk 14]
```diff
@@ -174,3 +175,31 @@ terminate_proto_reason(Reason) ->
         normal -> 'REASON_NORMAL_DISCONNECT';
         _ -> 'REASON_UNSPECIFIED'
     end.
+
+allowed_reasons() ->
```
[the remark this answers is not in the document]
Comment (VivekPipaliya23) [reply apps/vmq_server/src/vmq_mqtt_fsm_util.erl]: Removed

**Review by dhruvjain99 — CHANGES_REQUESTED** [review]

File: apps/vmq_server/src/vmq_mqtt_fsm_util.erl [hunk 15]
```diff
@@ -143,4 +144,33 @@ terminate_reason(?DISCONNECT_MIGRATION) -> normal;
...
+terminate_reason(?INVALID_PUBREC_ERROR) -> normal;
+terminate_reason(?INVALID_PUBCOMP_ERROR) -> normal;
+terminate_reason(?TCP_CLOSED) -> normal;
+terminate_reason(?EXIT_SIGNAL_RECEIVED) -> normal;
+terminate_reason(?PUBLISH_AUTH_ERROR) -> normal;
 terminate_reason(Reason) -> Reason.
+
+-spec terminate_proto_reason(any()) -> any().
+terminate_proto_reason(Reason) ->
+    case Reason of
```
[the remark this answers is not in the document]
Comment (VivekPipaliya23) [reply apps/vmq_server/src/vmq_mqtt_fsm_util.erl:157]: done

**Comment by dhruvjain99** [comment]
@VivekPipaliya23 I can see crashed processes in the tests ci. Can you please fix it?

**Comment by VivekPipaliya23** [comment]
> @VivekPipaliya23 I can see crashed processes in the tests ci. Can you please fix it?

I have added a commit fixing that.

File: apps/vmq_commons/src/vmq_types_mqtt5.hrl [hunk 16]
```diff
@@ -48,6 +48,7 @@
 -define(M5_MAX_CONNECT_TIME, 16#A0).
 -define(M5_SUBSCRIPTION_IDS_NOT_SUPPORTED, 16#A1).
 -define(M5_WILDCARD_SUBS_NOT_SUPPORTED, 16#A2).
+-define(M5_TCP_CLOSED, 16#A3).
```
Comment (dhruvjain99) [inline apps/vmq_commons/src/vmq_types_mqtt5.hrl:51]: What is 16#A3?

**Review by dhruvjain99 — CHANGES_REQUESTED** [review]

File: apps/vmq_commons/src/vmq_types_mqtt5.hrl (hunk 16 above)
Comment (dhruvjain99) [reply apps/vmq_commons/src/vmq_types_mqtt5.hrl:51]: Let's not do any kind of protocol change for this feature.

Comment (VivekPipaliya23) [reply apps/vmq_commons/src/vmq_types_mqtt5.hrl]: Removed

**Review by dhruvjain99 — APPROVED** [review]
LGTM!

**Pull request merged**